docs/type-annotations.org

```diff
@@ -0,0 +1,219 @@
+  hold anything, write =(list mixed)= or simply =list= for short.
+- =(vector a)= where =a= is the type of items in the vector.  If the
+  vector can hold anything, write =(vector mixed)= or simply =vector= for
+  short.
+- =(hash-table k v)= where =k= is the key type and =v= is the value type.
+  If the hash table can hold anything, write =(hash-table mixed mixed)=
+  or simply =hash-table= for short.
+
+** Function types
+
+Function types are types of functions.  They have input argument types
+and a return type.
+
+The function =add-one= from the introduction has a function type =(function
+(int) int)= which means it takes in one integer and returns an integer.
+
+A =lambda= form =(lambda (x) (number-to-string x))= has function type
+=(function (number) string)=, it takes in a number and returns a string.
+
+A function can have a function type as one of its input types.  An
+example of such a function is =mapcar= which takes a function and a list
+and applies the function to every item of the list.
+
+#+BEGIN_SRC elisp
+(defun app (fn)
+  "Apply FN to the list (1 2 3 4)"
+  (declare (elsa ((function (number) number)) (list number)))
+  (mapcar fn (list 1 2 3 4)))
+
+(app (lambda (x) (* x x)))
+#+END_SRC
+
+The =app= function requires that we pass in a function which processes a
+number into a number and returns a list of numbers.
+
+** Generic types
+
+Generic types are types where some of the type arguments are variable.
+Both basic and composite types can be turned into generic types.
+
+*** Motivation
+
+An example of a generic function is =identity=.  This function takes
+anything in and anything out.  We could therefore give it a type
+annotation =(elsa (mixed) mixed)=.
+
+However, we can do better!  We know that whatever was passed in will
+be returned and so the type actually must be the same.  The =(elsa
+(mixed) mixed)= signature allows us to pass in an =int= and it can return
+back a =string= no problem and so it would not catch a huge number of
+possible errors.
+
+What we want to express here is "X comes in, X comes out".
+
+*** Syntax
+
+The syntax for generic types is "generic type name" + =*= suffix.  Any
+string can be used for the generic type name, but customarily
+single-letter names are used.
+
+For the above mentioned identity function we therefore write the type
+as =(elsa (a*) a*)= where =a*= stands for a generic type =a=.
+
+A function such as =car= can be typed as follows:
+
+#+BEGIN_SRC elisp
+(elsa ((cons a* b*)) a*)
+#+END_SRC
+
+It takes a cons with =a= in the =car= and =b= in the =cdr= and return the =car=
+which is of type =a= , whatever that happens to be.
+
+** Optional types
+
+If a function can take optional arguments we preceed them with
+=&optional= marker just as we do in the argument list.
+
+#+BEGIN_SRC elisp
+(defun drop-items (list &optional n)
+  "Drop first item of LIST or N items if N is provided."
+  (declare (elsa ((list a*) &optional int) (list a*)))
```

Comment:
The `&optional` marker is not actually necessary since we can get that info from the function signature itself.  I feel it might be a good idea to have it redundant for cross-checking, but it might prove more annoying than useful.

Same for `&rest`

Reply:
With `&rest` there's also the issue that we might want to wrap the type with `(list)` or assume that implicitly.

- `(elsa (&rest string) string` implicit list
- `(elsa (&rest (list string)) string`
- `(elsa (string) string)` if we decide to get rid of `&rest` + implicit list
- `(elsa ((list string)) string)` if we decide to get rid of `&rest`

Reply:
Why not just* make `&optional` and `&rest` optional? The extent to which they help with readability rather depends on the argument list at hand. For the simplest functions, they don't add much: something like
```elisp
(defun make-csv-row (&rest strings) 
  (declare (elsa (string) string))
  (string-join strings ","))
```
reads perfectly naturally without a second `&rest`, but the redundant annotations become useful very quickly as the argument list's complexity grows. I would think providing them should be considered good style, in general.

​* "just" in the sense that "both!" is a _simple_ answer to either/or questions; I don't have a clear sense of what the implementation difficulty would actually be.

Reply:
@ambirdsall I kind of like this.  I'm a bit afraid people will try to derive some extra significance out of it though. i.e. lack of `&rest` meaning something as opposed to being optional.  Documentation helps of course but it's still unobvious at first.

But I think we can make it optional and see the reactions.  We can always deprecate that usage and ask users to fill in missing types (e.g. Elsa itself can check this and provide an error)